packages/dd-trace/src/profiling/profilers/wall.js

```diff
@@ -20,7 +54,63 @@ class NativeWallProfiler {
       process._stopProfilerIdleNotifier = () => {}
     }
 
+    this.resetStack()
     this._record()
+    if (this._hotspots) {
+      beforeCh.subscribe(this._enter)
+      afterCh.subscribe(this._exit)
+      incomingHttpRequestStart.subscribe(this._enter)
+      incomingHttpRequestEnd.subscribe(this._exit)
+    }
+  }
+
+  markAsSampled (span) {
+    // NOTE: there's no guarantee these tags will be applied to the span as it
+    // is possible it'll be sent to the agent by the time we get to execute
+    // this code.
+    if (span && this._labelsCaptured()) {
+      span.setTag('sampled', 'yes')
+      span.setTag('manual.keep', true)
+    }
+  }
+
+  setLabels (labels) {
+    this._currentLabels = labels
+    this._setLabels(labels)
+  }
+
+  _enter () {
+    if (!this._setLabels) return
+
+    const lastSpan = this._currentSpan
+    this.markAsSampled(lastSpan)
+    // NOTE: We stack nulls/undefineds *except* at the bottom of the
+    // stack, since pop() on an empty stack "synthesizes" those anyway.
+    if (lastSpan || this._spanStack.length > 0) {
+      this._spanStack.push(lastSpan)
+    }
+
+    const currentSpan = getActiveSpan() || null
+    this._currentSpan = currentSpan
+
+    const activeCtx = currentSpan ? currentSpan.context() : null
+
+    const labels = activeCtx ? {
+      'span id': activeCtx.toSpanId()
+    } : null
+
+    if (this._currentLabels || this._labelStack.length > 0) {
+      this._labelStack.push(this._currentLabels)
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**🎯 suggestion:** ‏ using a single stack for both spans and labels would ensure that both stacks won't go out of sync.

Reply:
I wanted to avoid creation of an intermediate object to contain the tuple. This code is somewhat performance critical, so I try to keep it as light as possible.

Reply:
I removed the span stack entirely, we only used it for `markAsSampled` but that's not terribly useful.